### ISSUE TITLE
IPLD requirements for Filecoin

### ISSUE THREAD

**Issue by mikeal** [filecoin-project/go-filecoin#2079]
Hey Filecoin folks :)

The IPLD team is trying to get a handle on our roadmap and plan for the next few quarters. Filecoin is a very important consumer of IPLD and we’d like to get a good list of things Filecoin needs from over the next year or so, along with priority and timeline.

Please list what you need along with any planned ship dates that would be blocked by not having it :)

**Comment by rosalinekarr** [comment]
@mikeal Would you mind if we do this in an issue on the IPLD project instead of here? We're already dealing with a pretty huge issue list, and I'd like to make sure this doesn't get lost.

**Comment by mishmosh** [comment]
Thanks @mikeal for kicking off this conversation. I don't mind keeping it in here since that's where it was started. 

We're still playing catchup from last week and don't have a forward-thinking list yet. Is next Thurs too late to get you that list?

In the short term, take a look at issues in `go-filecoin` mentioning ipld: https://github.com/filecoin-project/go-filecoin/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+ipld

* Most of them are about how filecoin _uses_ ipld but there are a few patterns worth thinking about. 
* There is a swirl of inter-related issues on capabilities of `go-ipld-cbor` (start here:
https://github.com/filecoin-project/go-filecoin/issues/643)
* Also, this specific issue seems stalled:
https://github.com/filecoin-project/go-filecoin/issues/626

**Comment by mikeal** [comment]
Next Thursday is fine, this isn’t urgent, we just want to start this conversation long enough before the end of the quarter that we can plan accordingly :)

**Comment by mishmosh** [comment]
Hi IPLD folks :) From our perspective, the reformat library is the primary link between filecoin <> ipld. Our ask is centered around making sure the reformat library is watertight ☔️.  We think it could take shape in several stages:

**1. Self-assessment from IPLD team on the reformat library.** What does currently do well & reliably? Where could it improve? Do you feel there are any key missing pieces or weaknesses?

**2. A well-tested reformat library.** Some kind of manual and/or automated process to assure reliability as features evolve. 

**3. Bugfixes & small feature improvements for specific issues - To date, we've found a couple small bugs; @phritz will elaborate further

Note this is very similar to the ask we're making of libp2p and relay dependencies.

**Comment by mikeal** [comment]
@whyrusleeping recently @warpfork mentioned that you had some concerns about cbor performance. Is this something you’d want the IPLD team to handle or is this something you’re planning on resourcing in Filecoin?

**Comment by warpfork** [comment]
A big picture piece of context to make sure we've got in the conversation: in IPLD land, we're also working on `go-ipld-prime` now, and long-run that might become the face of that "primary link".  I probably won't successfully describe all that that implies in one comment, but just to get things primed for further conversation:

Currently Filecoin is using `go-ipld-cbor`, which internally shells out to (a lot) of `refmt` code, and the way that integration with native golang types is done means lots of refmt "`Atlas`" objects are written in the Filecoin code to describe how to serialize things.  Refmt *only* understands this model of having developers write their own golang types and then write binding definitions in the form of Atlases.

In the future, `go-ipld-prime` is planned to offer a replacement to `go-ipld-cbor`, and addresses several significant architectural issues.  Some of these significant architectural issues regard performance: `go-ipld-cbor` tends to cause *numerous* memcopies between serial, "Node", and native-typed variants of the date (and there's several issues in filecoin that can trace the roots of performance issues to this); `go-ipld-prime` is designed from the start to make sure this can be avoided in almost all common cases. Another key feature is that will be other ways to define objects than just refmt Atlases: one planned feature is similar to the Atlas system for binding existing types; another planned feature is to use a Schema description, and have `go-ipld-prime` generate *both* the native golang types and all the serialization mappings at the same time, which will be able to bring significant benefits and developer-time-savings.

---

We're still working on `go-ipld-prime`, but the intent is to get it ready to go "soon" with at least minimum viable  features; and follow up with that schema and codegen work as fast as time allows.

`refmt` is getting a lower priority by sheer lack of developer hours.  `refmt` isn't abandonware, by any means (and in particular, we're still *absolutely* using a bunch of the cbor parser stuff in the new `go-ipld-prime` code, so improvements there will retain their value no matter what), but we do want to prioritize our time such that `go-ipld-prime` successfully ships sooner than later, because some of the major architectural issues with the present-tense `go-ipld-cbor` are not going to be solved by anything less.

**Comment by whyrusleeping** [comment]
@mishmosh on refmt specifically, it tends to not actually be our problem. The bugs we experienced were because, at the time, we hadnt updated the version of refmt we were using for around 8 months. 

The primary IPLD problem we have right now is that the go-ipld-cbor library (that I wrote) sucks. It copies things too much and overallocates, and does way more work than it should have to. Additionally, its not that nice to use. go-ipld-prime is meant to replace that mess for us.

Some of the work I did a month or two ago around how blocks get serialized and how their Cids are computed helped the problem noticeably (there was some code abuse going on, doing things way more times than was needed). There may be other low hanging fruit in gfc we can do here to improve the state of things.
